core/deployment/src/main/java/io/quarkus/deployment/dev/JavaCompilationProvider.java

```diff
@@ -49,57 +56,64 @@ public Set<String> handledExtensions() {
+                    context.getReleaseJavaVersion(),
+                    context.getSourceJavaVersion(),
+                    context.getTargetJvmVersion()).toList();
         }
+
+        final JavaCompiler compiler = this.compiler;
+
         if (compiler == null) {
             throw new RuntimeException("No system java compiler provided");
         }
-        try {
-            if (fileManager == null) {
-                fileManager = compiler.getStandardFileManager(fileManagerDiagnostics = new DiagnosticCollector<>(), null,
-                        context.getSourceEncoding());
-            }
 
-            DiagnosticCollector<JavaFileObject> diagnostics = new DiagnosticCollector<>();
-            fileManager.setLocation(StandardLocation.CLASS_PATH, context.getClasspath());
-            fileManager.setLocation(StandardLocation.CLASS_OUTPUT, List.of(context.getOutputDirectory()));
+        final QuarkusFileManager.Context sourcesContext = new QuarkusFileManager.Context(
+                context.getClasspath(), context.getReloadableClasspath(),
+                context.getOutputDirectory(), context.getSourceEncoding());
 
-            CompilerFlags compilerFlags = new CompilerFlags(COMPILER_OPTIONS, context.getCompilerOptions(getProviderKey()),
-                    context.getReleaseJavaVersion(), context.getSourceJavaVersion(), context.getTargetJvmVersion());
+        if (this.fileManager == null) {
+            final Supplier<StandardJavaFileManager> supplier = () -> {
+                final Charset charset = context.getSourceEncoding();
+                return compiler.getStandardFileManager(null, null, charset);
+            };
 
-            Iterable<? extends JavaFileObject> sources = fileManager.getJavaFileObjectsFromFiles(filesToCompile);
-            JavaCompiler.CompilationTask task = compiler.getTask(null, fileManager, diagnostics,
-                    compilerFlags.toList(), null, sources);
-
-            if (!task.call()) {
-                StringBuilder sb = new StringBuilder("\u001B[91mCompilation Failed:");
-                for (Diagnostic<? extends JavaFileObject> i : diagnostics.getDiagnostics()) {
-                    sb.append("\n");
-                    sb.append(i.toString());
-                }
-                sb.append("\u001b[0m");
-                throw new RuntimeException(sb.toString());
+            if (context.getReloadableClasspath().isEmpty()) {
+                this.fileManager = new StaticFileManager(supplier, sourcesContext);
+            } else {
+                this.fileManager = new ReloadableFileManager(supplier, sourcesContext);
             }
+        } else {
+            this.fileManager.reset(sourcesContext);
+        }
 
-            logDiagnostics(diagnostics);
+        final DiagnosticCollector<JavaFileObject> diagnosticsCollector = new DiagnosticCollector<>();
 
-            if (!fileManagerDiagnostics.getDiagnostics().isEmpty()) {
```

Comment:
There is no need to keep a diagnostic collector for the file manager, only the compiler diagnostics is useful for troubleshooting